src/wifi_dpp.rs

```diff
@@ -0,0 +1,451 @@
+                Err(EspError::from_infallible::<ESP_ERR_INVALID_STATE>())
+            }
+            other => {
+                self.stop_listen();
+                Err(unexpected_state(other))
+            }
+        }
+    }
+
+    /// Stop listening for credentials.  If any callers are actively blocked waiting for credentials
+    /// they will be notified with EspError(ESP_ERR_INVALID_STATE).  This method is not
+    /// necessary to call after [Self::wait_for_credentials] or
+    /// [Self::wait_for_credentials_with_timeout] returns as the esp_dpp API automatically
+    /// stops listening on success or failure.
+    pub fn stop_listen(&self) {
+        info!("Stopping DPP listener...");
+
+        // SAFETY: This function should be safe to call without any locks as it mostly just
+        // sets a flag to halt at next chance.
+        unsafe { esp_supp_dpp_stop_listen() };
+
+        self.bootstrapped.dpp.set_pending_state(DppState::Stopped);
+    }
+
+    /// Attempt to retry after a transient failure in [Self::wait_for_credentials] (for example
+    /// if a user scanned a bogus QR code or there was a recoverable transmit error on the
+    /// channel).  This API will fail at runtime if ESP-IDF is not patched with
+    /// https://github.com/espressif/esp-idf/pull/10865.
+    ///
+    /// On failure to retry, the caller is expected to bootstrap again in order to logically retry
+    /// however this will potentially lose state and generate a new QR code if any parameters
+    /// change (such as a new key being generated).
+    pub fn attempt_retry(self) -> Result<EspWifiDpp<'d, 'w, T>, ()> {
+        if Self::is_start_listen_patched() {
+            Ok(self.bootstrapped)
+        } else {
+            Err(())
+        }
+    }
+
+    // TODO: Sure would be nice to be able to write esp_idf_version >= 5.1...
+    #[cfg(any(
+        esp_idf_version_major = "4",
+        all(esp_idf_version_major = "5", esp_idf_version_minor = "0")
+    ))]
+    fn is_start_listen_patched() -> bool {
+        false
+    }
+
+    #[cfg(not(any(
+        esp_idf_version_major = "4",
+        all(esp_idf_version_major = "5", esp_idf_version_minor = "0")
+    )))]
+    fn is_start_listen_patched() -> bool {
+        true
+    }
+}
+
+fn unexpected_state(state: DppState) -> EspError {
+    warn!("Unexpected DPP state: {state:?}");
+    EspError::from_infallible::<ESP_ERR_INVALID_STATE>()
+}
+
+#[derive(Debug)]
+enum DppState {
+    BootstrappedUriReady(String),
+    ConfigurationReceived(ClientConfiguration),
+    Fail(EspError),
+    Stopped,
+}
+
+pub struct QrCode(pub String);
+
+#[cfg(esp_idf_version_major = "4")]
+/// ESP-IDF 4.x internally framed the key inside esp_dpp APIs.
+fn frame_key(unframed: &[u8; 32]) -> &[u8] {
+    unframed
+}
+
+#[cfg(not(esp_idf_version_major = "4"))]
+/// ESP-IDF 5.x requires the caller put the key into the PEM format
+fn frame_key(unframed: &[u8; 32]) -> Vec<u8> {
```

Comment:
Why calling into `alloc` just for this? You can either return a `heapless::Vec` (as the size of the returned thing is actually known), or even an array, as the final array size is also known?

Reply:
I was struggling with this one a bit going back and forth between a compile-time approach vs a runtime one (compile-time can do as you say with heapless, runtime would need alloc since you wouldn't know the result size).  Since I settled on compile-time, I'll go ahead and slip in `heapless::Vec` again.